ddtrace/internal/utils/attrdict.py

```diff
@@ -20,23 +19,39 @@ class AttrDict(dict):
        print(data.key)
     """
 
-    def __getattr__(self, key):
+    def __init__(self, *args, **kwargs):
+        # type: (Any, Any) -> None
+        self.__dict__.update(*args, **kwargs)
+
+    def __getattr__(self, name):
+        # type: (str) -> Any
+        return getattr(self.__dict__, name)
+
+    def __contains__(self, name):
+        # type: (str) -> bool
+        return name in self.__dict__
+
+    def __getitem__(self, name):
         # type: (str) -> Any
-        if key in self:
-            return self[key]
-        return object.__getattribute__(self, key)
+        return self.__dict__[name]
 
-    def __setattr__(self, key, value):
+    def __setitem__(self, name, value):
         # type: (str, Any) -> None
-        # 1) Ensure if the key exists from a dict key we always prefer that
-        # 2) If we do not have an existing key but we do have an attr, set that
-        # 3) No existing key or attr exists, so set a key
-        if key in self:
-            # Update any existing key
-            self[key] = value
-        elif hasattr(self, key):
-            # Allow overwriting an existing attribute, e.g. `self.global_config = dict()`
-            object.__setattr__(self, key, value)
-        else:
-            # Set a new key
-            self[key] = value
+        self.__dict__[name] = value
+
+
+class DefaultAttrDict(AttrDict):
+    """AttrDict with a default value for missing keys."""
+
+    def __init__(self, default):
+        # type: (Callable[[DefaultAttrDict, str], Any]) -> None
+        super(DefaultAttrDict, self).__init__()
+        self.__default_attrdict_constructor__ = default
+
+    def __getattr__(self, name):
```

Comment:
Could you give an example of how this provides a default value for missing keys? I'm not sure I understand what `self.__default_attrdict_constructor__(self, name)` does and how it provides an analogue of `defaultdict`.

Reply:
~~~ python
>>> default_config = DefaultAttrDict(IntegrationConfig)
>>> default_config.some_integration
IntegrationConfig("some_integration")
~~~
This is based on the fact that `__getattr__` is called _conditionally_ when the attribute is not actually present on the instance (`default_config` in the example below). In this case we first try bubbling up the request for the attribute to the superclass(es). If we still can't find an attribute, we create one using the constructor we passed to `DefaultAttrDict` (`IntegrationConfig` in this case, which takes two arguments), store it as an item (which can be accessed as an attribute because of the super `AttrDict`), and return that. Subsequent attempts to access the same attribute (`some_integration` in this example), will return the instance of `IntegrationConfig` created on the first access.